infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -30,96 +33,197 @@ private StateTransfer() {
-                    // Since the LU will reject trim decisions made from older epochs, we
-                    // need to adjust the new trim mark to have the new layout's epoch.
-                    Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
-                    return runtime.getLayoutView().getRuntimeLayout(layout)
-                            .getLogUnitClient(endpoint)
-                            .prefixTrim(prefixToken);
-                })
-                .forEach(CFUtils::getUninterruptibly);
-
-        if (trimMark > segment.getEnd()) {
-            log.info("stateTransfer: Nothing to transfer, trimMark {} greater than end of segment {}",
-                    trimMark, segment.getEnd());
-            return;
-        }
+            try {
 
-        // State transfer should start from segment start address or trim mark whichever is lower.
-        long segmentStart = Math.max(trimMark, segment.getStart());
+                long trimMark = setTrimOnNewLogUnit(layout, runtime, endpoint);
 
-        for (long chunkStart = segmentStart; chunkStart < segment.getEnd()
-                ; chunkStart = chunkStart + batchSize) {
-            long chunkEnd = Math.min((chunkStart + batchSize - 1), segment.getEnd() - 1);
+                if (trimMark > segment.getEnd()) {
+                    log.info("stateTransfer: Nothing to transfer, trimMark {}"
+                                    + "greater than end of segment {}",
+                            trimMark, segment.getEnd());
+                    return true;
+                }
 
-            long ts1 = System.currentTimeMillis();
+                // State transfer should start from segment start address or trim mark
+                // whichever is higher.
+                final long segmentStart = Math.max(trimMark, segment.getStart());
+                final long segmentEnd = segment.getEnd() - 1;
+                log.info("stateTransfer: Total address range to transfer: [{}-{}] to node {}",
+                        segmentStart, segmentEnd, endpoint);
 
-            Map<Long, ILogData> dataMap = runtime.getAddressSpaceView()
-                    .fetchAll(ContiguousSet.create(Range.closed(chunkStart, chunkEnd),
-                            DiscreteDomain.longs()), true);
+                // Transfer non-written chunks of addresses.
+                for (long chunkStart = segmentStart; chunkStart <= segmentEnd
+                        ; chunkStart += chunkSize) {
 
-            long ts2 = System.currentTimeMillis();
+                    long chunkEnd = Math.min(segmentEnd, chunkStart + chunkSize - 1);
 
-            log.info("stateTransfer: read {}-{} in {} ms", chunkStart, chunkEnd, (ts2 - ts1));
+                    // Fetch all missing entries in this range [chunkStart - chunkEnd].
+                    List<Long> chunk = getMissingEntriesChunk(layout, runtime, endpoint,
+                            chunkStart, chunkEnd);
+                    if (chunk.isEmpty()) {
```

Comment:
Please log that chunk is being omitted because the orchestrator didn't detect it as missing

Reply:
Since `transferChunk` accesses the first and last index of the list, it assumes that the list is not empty.
It would make sense to move this check in side  the `transferChunk` method.

Reply:
Moved the check to transferChunk.
That is already being logged in getMissingEntries.